SU2_CFD/src/solvers/CFEASolver.cpp

```diff
@@ -0,0 +1,3774 @@
+  Set_ElementProperties(geometry, config);
+
+  Total_CFEA        = 0.0;
+  WAitken_Dyn       = 0.0;
+  WAitken_Dyn_tn1   = 0.0;
+  loadIncrement     = 0.0;
+
+  SetFSI_ConvValue(0,0.0);
+  SetFSI_ConvValue(1,0.0);
+
+  nVar = nDim;
+
+  /*--- Define some auxiliary vectors related to the residual ---*/
+
+  Residual_RMS = new su2double[nVar]();
+  Residual_Max = new su2double[nVar]();
+  Point_Max = new unsigned long[nVar]();
+  Point_Max_Coord = new su2double*[nVar];
+  for (iVar = 0; iVar < nVar; iVar++) {
+    Point_Max_Coord[iVar] = new su2double[nDim]();
+  }
+
+  /*--- The length of the solution vector depends on whether the problem is static or dynamic ---*/
+
+  unsigned short nSolVar;
+  string text_line, filename;
+  ifstream restart_file;
+
+  if (dynamic) nSolVar = 3 * nVar;
+  else nSolVar = nVar;
+
+  su2double* SolInit = new su2double[nSolVar]();
+
+  /*--- Initialize from zero everywhere ---*/
+
+  nodes = new CFEABoundVariable(SolInit, nPoint, nDim, nVar, config);
+  SetBaseClassPointerToNodes();
+
+  delete [] SolInit;
+
+  /*--- Set which points are vertices and allocate boundary data. ---*/
+
+  for (iPoint = 0; iPoint < nPoint; iPoint++)
+    for (unsigned short iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++) {
+      long iVertex = geometry->node[iPoint]->GetVertex(iMarker);
+      if (iVertex >= 0) {
+        nodes->Set_isVertex(iPoint,true);
+        break;
+      }
+    }
+  static_cast<CFEABoundVariable*>(nodes)->AllocateBoundaryVariables(config);
+
+
+  if (config->GetRefGeom()) Set_ReferenceGeometry(geometry, config);
+  if (config->GetPrestretch()) Set_Prestretch(geometry, config);
+
+  /*--- Initialization of matrix structures ---*/
+  if (rank == MASTER_NODE) cout << "Initialize Jacobian structure (Non-Linear Elasticity)." << endl;
+
+  Jacobian.Initialize(nPoint, nPointDomain, nVar, nVar, false, geometry, config);
+
+  if (dynamic) {
+    MassMatrix.Initialize(nPoint, nPointDomain, nVar, nVar, false, geometry, config);
+    TimeRes_Aux.Initialize(nPoint, nPointDomain, nVar, 0.0);
+    TimeRes.Initialize(nPoint, nPointDomain, nVar, 0.0);
+  }
+
+  /*--- Initialization of linear solver structures ---*/
+  LinSysSol.Initialize(nPoint, nPointDomain, nVar, 0.0);
+  LinSysRes.Initialize(nPoint, nPointDomain, nVar, 0.0);
+
+  LinSysAux.Initialize(nPoint, nPointDomain, nVar, 0.0);
+
+  LinSysReact.Initialize(nPoint, nPointDomain, nVar, 0.0);
+
+#ifdef HAVE_OMP
+  /*--- Get the element coloring. ---*/
+
+  const auto& coloring = geometry->GetElementColoring();
+
+  auto nColor = coloring.getOuterSize();
+  ElemColoring.resize(nColor);
+
+  for(auto iColor = 0ul; iColor < nColor; ++iColor) {
+    ElemColoring[iColor].size = coloring.getNumNonZeros(iColor);
+    ElemColoring[iColor].indices = coloring.innerIdx(iColor);
+  }
+
+  ColorGroupSize = 1; /// TODO: This needs to come from geometry or config
+
+  omp_chunk_size = computeStaticChunkSize(nPointDomain, omp_get_max_threads(), OMP_MAX_SIZE);
+#endif
```

Comment:
The element coloring is obtained from CGeometry as a shallow copy at construction.
I have yet to test the performance impact of the color group size, group size 1 gives maximum parallelization but worse spatial locality.

Reply:
Update: The impact of not grouping elements for coloring (size 1) is not significant for the FEA solver, not a lot of data is accessed per element relative to the computation effort and linear algebra takes up 80%+ of the time anyway.